allow syncing single users, add verbose and very verbose output

butonic opened the pull request:
fixes https://github.com/owncloud/core/issues/27707

butonic committed:
allow syncing single users, add verbose and very verbose output

tomneedham commented:
We should get this in 10.0.3 to help big instances sync specific users + aid scripting

jvillafanez commented:
I'll have to review this once this is ready. There are some code pieces misplaced

PVince81 commented:
moving to "triage" for later rescheduling

PVince81 commented:
unfinished, moving to "planned"

butonic committed:
fix-tests

PVince81 commented:
please post the test plan that you used when testing this

jvillafanez commented on lib/private/User/SyncService.php:
```diff
@@ -89,40 +92,79 @@ public function getNoLongerExistingUsers(\Closure $callback) {
+	 * @param int $maxUsers for the progress bar
 	 */
-	public function run(\Closure $callback) {
+	public function run(OutputInterface $output, $maxUsers = 0) {
```
This isn't the place to use symfony's OutputInterface nor ProgressBar. This class shouldn't know how to write in the console, and it should be limited to call the callback for each user.
The callback approach is a better choice in this case.

jvillafanez commented on lib/private/Migration/OutputAdapter.php:
```diff
@@ -0,0 +1,147 @@
+use Symfony\Component\Console\Formatter\OutputFormatterInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class OutputAdapter implements OutputInterface {
```
I'll need someone to convince me this is a good idea.

I think @DeepDiver1975 already created a ConsoleLogger somewhere to provide a log interface connected to a symfony console output. It might be better to enrich that.

jvillafanez replied:
Maybe I got confused with https://github.com/owncloud/core/blob/master/lib/private/Migration/ConsoleOutput.php which is similar to what I had in mind...

I'll recheck this

jvillafanez commented:
My main concern is that we should bring back the callback approach to the sync service. The sync service should just focus on syncing, not writing; the callback will take care of writing whatever is needed, or move the progress bar, or send an email to the synced user.

If you don't like callbacks, you can create a custom interface for the sync service, so the sync service can call whatever method of the interface.
```
use \OCP\IUser;
public interface ISyncServiceCallback {
    public function callMeWith(IUser $user);
}
```
Whatever happens there isn't sync service's bussines.

jvillafanez commented:
The reason to use the ILogger is to provide a common logging interface across ownCloud. No app shouldn't care about where or how the log will end. If the sync service needs to log something, an instance of the ILogger should be injected.

Now, we usually use the ILogger set up by the ownCloud server, BUT it doesn't implies that we can't set up other ILoggers in different scenarios. Custom implementations can be created. An example could be https://github.com/owncloud/windows_network_drive/blob/master/lib/custom_loggers/consolelogger.php which is in use for console commands.
Obviously, the specific set up for the ILogger will depend on the specific implementation: if the implementation needs a IConfig object to get the minimum log level, this can be set up before injecting the logger implementation.

As said, I don't see a reason to use symfony classes directly in the sync service (at least console related classes, other classes might have sense). It's fine if they are used and / or wrapped in the sync command, but not below that.

The reason to use a callback (or something similar) isn't for logging, it's for notifying events to the caller, in this particular case that the specified user has been synced. If that event is only used for logging, it doesn't matter for the sync service. It can also be used to send a welcome email or something similar, but as said it isn't sync service's business.
It might have more sense to use a callback interface instead of a callback or a closure in case we want to include more things there: call for new users, updated user, removed users...

butonic committed:
use callback class to handle output

butonic commented:
@jvillafanez thx for rubbing this in ... my OOD foo seems to have deteriorated. I have been away from Java too long. I did not use an Interface, yet. The class can be mocked just as well and can be introduced if someone really needs it.

Anyway, I also added a nowline flag to the IOutput intefaces info and warning methods so the console can start a Sync with `Syncing <userid>: ` and end with `updated in 0.02...s` in verbose output mode.

The `Analyse unknown users ...` is now `Analysing synced users ...` and shows the usercount from the account table, which is a cheap db query, in contrast to the count for users on an ldap backend. That is why I omit counting the users when not in normal verbosity mode.

@PVince81 to test just sync your ldap as usual.
0. normal verbosity only changes the `Analysing ...` string, as mentioned above.
1. When using -v you should see
```
$ occ user:sync "OCA\User_LDAP\User_Proxy" -v -m disable
Insert new and update existing users ...
Analysing synced users ...
 2843/2843 [============================] 100% 26 secs

No unknown users have been detected.
Syncing Nadja_Storm: updated in 0.027519941329956s
Syncing Susan_Junge: updated in 0.021672010421753s
...
```
2. with -vv it becomes
```
$ occ user:sync "OCA\User_LDAP\User_Proxy" -vv -m disable
Insert new and update existing users ...
Analysing synced users ...
 2843/2843 [============================] 100% 25 secs/25 secs

No unknown users have been detected.
Syncing Nadja_Storm: updated with {"id":52733,"backend":"OCA\\User_LDAP\\User_Proxy","userId":"Nadja_Storm","displayName":"Storm, Nadja","email":"zombie295@example.org","home":"\/mnt\/c\/Users\/jfd\/Repositories\/oc\/core\/data\/Nadja_Storm","lastLogin":0,"quota":"default","searchTerms":["zombie295@example.org"]} in 0.026507139205933s
Syncing Susan_Junge: updated with {"id":52734,"backend":"OCA\\User_LDAP\\User_Proxy","userId":"Susan_Junge","displayName":"Junge, Susan","email":"zombie296@example.org","home":"\/mnt\/c\/Users\/jfd\/Repositories\/oc\/core\/data\/Susan_Junge","lastLogin":0,"quota":"default","searchTerms":["zombie296@example.org"]} in 0.022819042205811s
...
```
3. Using -vvv currently gives the same output, but in the future should also show messages from LDAP und SQL queries. At least IMNSHO.

Similar syncing single users with `-u` gives:
```
$ occ user:sync "OCA\User_LDAP\User_Proxy" -m disable -u Hans-Christian_Krauss
Insert new and update existing users ...
Analysing Hans-Christian_Krauss ...
Syncing Hans-Christian_Krauss: updated in 0.012651205062866s


$ occ user:sync "OCA\User_LDAP\User_Proxy" -v -m disable -u Hans-Christian_Krauss
Insert new and update existing users ...
Analysing Hans-Christian_Krauss ...
Syncing Hans-Christian_Krauss: updated in 0.016822814941406s


$ occ user:sync "OCA\User_LDAP\User_Proxy" -vv -m disable -u Hans-Christian_Krauss
Insert new and update existing users ...
Analysing Hans-Christian_Krauss ...
Syncing Hans-Christian_Krauss: updated with {"id":52761,"backend":"OCA\\User_LDAP\\User_Proxy","userId":"Hans-Christian_Krauss","displayName":false,"email":null,"home":"\/mnt\/c\/Users\/jfd\/Repositories\/oc\/core\/data\/Hans-Christian_Krauss","lastLogin":0,"quota":null,"searchTerms":[]} in 0.014629125595093s


$ occ user:sync "OCA\User_LDAP\User_Proxy" -vvv -m disable -u Hans-Christian_Krauss
Insert new and update existing users ...
Analysing Hans-Christian_Krauss ...
Syncing Hans-Christian_Krauss: updated with {"id":52761,"backend":"OCA\\User_LDAP\\User_Proxy","userId":"Hans-Christian_Krauss","displayName":false,"email":null,"home":"\/mnt\/c\/Users\/jfd\/Repositories\/oc\/core\/data\/Hans-Christian_Krauss","lastLogin":0,"quota":null,"searchTerms":[]} in 0.015676021575928s
```

butonic committed:
fix typo

butonic committed:
remove lib/private/Migration/OutputAdapter.php, was a bad idea anyway

PVince81 commented:
please review, then backport ?

jvillafanez commented on lib/private/User/SyncServiceCallback.php:
```diff
@@ -0,0 +1,116 @@
+	 * TODO use ITimeFactory to test output when it provides microtime
+	 *
+	 * @param IOutput $output
+	 * @param int $logLevel
```
Need more information here. I'm not sure what is this log level used for without looking at the code, nor the valid values.

DeepDiver1975 commented:
too hugh change and really feels messy to use command interfaces outside of a command.
we did this once in encrpytion and this is honestly 💩 

lets try to find a way which fits better.

jvillafanez commented:
As long as the `SyncService` uses an `ISyncServiceCallback` interface, it should be fine. We'll need to figure out where we should put the implementation, but other than that this approach is fine from my point of view.

PVince81 commented:
@butonic see comments. Can this be finished before 10.0.4 RC1?
If not please defer to "planned"

DeepDiver1975 commented:
@ownclouders rebase

ownclouders commented:
Hey! This pull request is being rebased automatically. Please DO NOT push while rebase is in progress or your changes would be lost!

ownclouders commented:
Automated rebase was successful!

tomneedham commented:
Closing to stop people using this. We will reimplement this after https://github.com/owncloud/core/pull/29669 since things brings further code de-duplication and refactoring which is necessary.
